File: Background-OC/Background-OC/ViewController.m
//
//  ViewController.m
//  Background-OC
//
//  Created by David Yang on 15/3/24.
//  Copyright (c) 2015年 Sensoro. All rights reserved.
//

#import "ViewController.h"
#import <CoreLocation/CoreLocation.h>

@interface ViewController () <CLLocationManagerDelegate>{
    BOOL _started;
}

@property (nonatomic,strong) NSTimer * timer;
@property (nonatomic,strong) CLLocationManager * locationManager;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    _locationManager = [[CLLocationManager alloc] init];
    [_locationManager requestAlwaysAuthorization];
    _locationManager.delegate = self;
    
    _started = NO;
    
    _timer = [NSTimer scheduledTimerWithTimeInterval:5 target:self selector:@selector(logout:) userInfo:nil repeats:YES];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)actionPress:(id)sender {
    
    if (_started == NO) {
        _started = YES;
        [_locationManager startUpdatingLocation];
        
        [_actionButton setTitle:@"结束" forState:UIControlStateNormal];
        
    }else{
        _started = YES;
        [_locationManager stopUpdatingLocation];
        
        [_actionButton setTitle:@"开始" forState:UIControlStateNormal];
    }
    
}

- (void) logout : (NSTimer*) timer {

    static int timeCall = 0;
    NSLog(@"It is %d call!", timeCall++);
    
}

#pragma mark CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager *)manager
     didUpdateLocations:(NSArray *)locations{

    NSLog(@"did Update Locations");
}

@end

File: Background-Swift/Background-Swift/ViewController.swift
//
//  ViewController.swift
//  Background-Swift
//
//  Created by David Yang on 15/3/24.
//  Copyright (c) 2015年 Sensoro. All rights reserved.
//

import UIKit
import CoreLocation

class ViewController: UIViewController, CLLocationManagerDelegate{

    let locationManager = CLLocationManager();
    var started = false;
    var timer : Timer!;
    var timeCall = 0;
    
    @IBOutlet weak var actionButton: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        if #available(iOS 8.0, *) {
            locationManager.requestAlwaysAuthorization()
        } else {
            // Fallback on earlier versions
        };
        locationManager.delegate = self;
        //locationManager.pausesLocationUpdatesAutomatically = false; //打开此选项可以持续后台运行。关闭会导致一段时间后后台更新停止。
        
        timer = Timer.scheduledTimer(timeInterval: 5, target: self,
                                     selector: #selector(logout(timer:)), userInfo: nil, repeats: true);
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @IBAction func actionPress(sender: AnyObject) {
        if started == false{
            started = true;
            actionButton.setTitle("结束", for: .normal);
            
            locationManager.startUpdatingLocation();
        }else{
            started = false;
            actionButton.setTitle("开始", for: .normal);
            
            locationManager.stopUpdatingLocation();
        }
    }

    @objc func logout(timer : Timer) -> Void{
        NSLog("It is %d call!", timeCall);
        timeCall += 1;
    }
    
    //MARK: CLLocationManagerDelegate
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        NSLog("did Update Locations");
    }
}

